### PR TITLE
add(test): test disabled `lightwalletd` mempool gRPCs via zebrad logs

### LINKED ISSUE #4350: Add GetMempoolTx and GetMempoolStream gRPC tests
## Motivation

In https://github.com/ZcashFoundation/zebra/pull/4253 we added tests for grpc wallet methods, however we left a few TODOs in the main file (zebrad/tests/common/lightwalletd/wallet_grpc_test.rs) for missing tests. 

## Tasks

Add a test for:
- GetMempoolTx
- GetMempoolStream

The send transaction test populates the mempool, so we can add these extra gRPC calls to the end of that test.

## Draft Implementation

PR #4537 has a draft implementation.

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

We want to test the `lightwalletd` mempool gRPC queries, but they are currently disabled.

### Designs

Call the gRPC tests, but expect no results.
Check that `lightwalletd` is querying Zebra for the mempool transactions via the `zebrad` logs.

## Solution

- Test fetching sent mempool transactions using gRPC
- Add extra `zebrad` log checks to the send transaction test
- Check `zebrad` JSON-RPC logs instead of disabled `lightwalletd` gRPCs
- Add a debug option that makes RPCs pretend the sync is finished 

Extra logging:
- Add debugging output for mempool transaction verification

Related fixes:
- Only copy the inner state directory in the send transactions test
- Add transactions to the send transactions list, rather than replacing the list with each new block
- Preload Zcash parameters in some transaction verification tests
- Wait for zebrad mempool activation before running gRPC tests
- Add a block and transaction Hash method to convert from display order
- Update test coverage docs

Closes #4350 

## Review

@oxarbitrage is probably the best person to review this test, it uses a bunch of his code from PR #4989.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

If we're going to make more changes to the send transaction test, we might want to:
- #5015

### CODE REVIEW

**Comment by teor2345** [comment]
If you want to run this test locally, I'm using:

```sh
export TMPDIR=/home/dev/.cache/zebra-tmp
export ZEBRA_CACHED_STATE_DIR=/home/dev/.cache/zebra-custom
export LIGHTWALLETD_DATA_DIR=/home/dev/.cache/lightwalletd-custom
export RUST_LOG=info

cargo test --release \
      --test acceptance \
      --features lightwalletd-grpc-tests \
      -- \
      --nocapture --include-ignored \
      sending_transactions_using_lightwalletd
```

**Comment by teor2345** [comment]
That's weird, it looks like all the commits didn't get pushed to this PR. It compiled locally, I'll try to find them.

**Comment by oxarbitrage** [comment]
the command provided to run locally has 2 improvements that i was not using on my testing which are:
- `TMPDIR` env var, i didnt knew this was possible so i had the tmp folder hardcoded in the code to juse a different hard drive with enough space to create the copy.
- `--release `, this is great, i didnt realized you could do it for this tests, it speed up building time.

I think we should consider adding this to https://github.com/ZcashFoundation/zebra/blob/main/zebrad/tests/acceptance.rs#L96 but we can do that later.

**Comment by oxarbitrage** [comment]
> That's weird, it looks like all the commits didn't get pushed to this PR. It compiled locally, I'll try to find them.

It build locally as it is, it seems the problem is in the clippy.

**Comment by oxarbitrage** [comment]
I was running this test locally and it crashed after 2 hours, i think i might be too far away from the tip as it tried to sync and timed out:

```
2022-08-31T20:11:52.231739Z  INFO {zebrad="c1cb2d6" net="Main"}: zebrad::components::sync::progress: estimated progress to chain tip sync_percent=99.755% current_height=Height(1788271) network_upgrade=Nu5 remaining_sync_blocks=4401 time_since_last_state_block=0s
2022-08-31T20:12:50.216565Z  INFO {zebrad="c1cb2d6" net="Main"}:sync:try_to_sync: zebrad::components::sync: starting sync, obtaining new tips state_tip=Some(Height(1788271))
Error: 
   0: stdout of command did not log any matches for the given regex,
      within the FormattedDuration(3600s) command timeout

Location:
   /media/oxarbitrage/4eb53770-4e4f-4d80-a830-0914f3f5b89a/zebra/send-tx-grpc-mempool-tests/zebra/zebra-test/src/command.rs:856

Match Regex:
   [
       "activating mempool",
   ]

Command:
   "/media/oxarbitrage/4eb53770-4e4f-4d80-a830-0914f3f5b89a/zebra/send-tx-grpc-mempool-tests/zebra/target/release/zebrad" "-c" "/media/oxarbitrage/4eb53770-4e4f-4d80-a830-0914f3f5b89a/tmp/zebrad_testsqZ7F8s/zebrad.toml" "start"

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ SPANTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

   0: zebra_test::command::expect_line_matching_regexes with success_regexes=RegexSet(["activating mempool"]) stream_name="stdout"
      at zebra-test/src/command.rs:814
   1: zebra_test::command::expect_stdout_line_matches with success_regex="activating mempool"
      at zebra-test/src/command.rs:705
   2: acceptance::common::sync::sync_until with height=Height(2147483647) network=Mainnet stop_regex="finished initial sync to chain tip, using gossiped blocks .*sync_percent.*=.*100\\." timeout=3600s mempool_behavior=ShouldAutomaticallyActivate checkpoint_sync=true check_legacy_chain=false
      at zebrad/tests/common/sync.rs:181
   3: acceptance::common::sync::copy_state_and_perform_full_sync with network=Mainnet partial_sync_path="/media/oxarbitrage/4eb53770-4e4f-4d80-a830-0914f3f5b89a/chain/zebra"
      at zebrad/tests/common/sync.rs:302
   4: acceptance::common::lightwalletd::send_transaction_test::load_transactions_from_future_blocks with network=Mainnet zebrad_state_path="/media/oxarbitrage/4eb53770-4e4f-4d80-a830-0914f3f5b89a/chain/zebra"
      at zebrad/tests/common/lightwalletd/send_transaction_test.rs:248

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
---- end of test output ----
```

**Comment by teor2345** [comment]
> the command provided to run locally has 2 improvements that i was not using on my testing which are:
> 
>     TMPDIR env var, i didnt knew this was possible so i had the tmp folder hardcoded in the code to juse a different hard drive with enough space to create the copy.

I added this in this PR.

>     --release , this is great, i didnt realized you could do it for this tests, it speed up building time.

In ticket #4955 we're going to rename these tests, which would also be a good time to update those docs to use `entrypoint.sh`. That would automatically run with `--release` and any other improvements we add later.

> I was running this test locally and it crashed after 2 hours, i think i might be too far away from the tip as it tried to sync and timed out

I made all the sync timeouts the same in this PR, so that should fix it.

File: zebrad/tests/common/lightwalletd/send_transaction_test.rs [hunk 1]
```diff
@@ -107,32 +131,108 @@ pub async fn run() -> Result<()> {
...
+    tracing::info!("calling GetMempoolTx gRPC to fetch transactions...");
+    let mut transactions_stream = rpc_client
+        .get_mempool_tx(Exclude { txid: vec![] })
+        .await?
+        .into_inner();
+
+    zebrad.expect_stdout_line_matches("answered mempool request req=TransactionIds")?;
+
+    // GetMempoolTx: make sure at least one of the transactions were inserted into the mempool.
+    let mut counter = 0;
```
Comment (teor2345) [inline zebrad/tests/common/lightwalletd/send_transaction_test.rs:191]:
```suggestion
    tracing::info!("calling GetMempoolTx gRPC to fetch transactions...");
    let mut counter = 0;
```

File: zebrad/tests/common/lightwalletd/send_transaction_test.rs [hunk 2]
```diff
@@ -107,32 +131,108 @@ pub async fn run() -> Result<()> {
...
+    tracing::info!("waiting for mempool to verify some transactions...");
+    zebrad.expect_stdout_line_matches("sending mempool transaction broadcast")?;
+
+    tracing::info!("calling GetMempoolTx gRPC to fetch transactions...");
+    let mut transactions_stream = rpc_client
+        .get_mempool_tx(Exclude { txid: vec![] })
+        .await?
+        .into_inner();
+
+    zebrad.expect_stdout_line_matches("answered mempool request req=TransactionIds")?;
```
Comment (teor2345) [inline zebrad/tests/common/lightwalletd/send_transaction_test.rs]:
```suggestion
    // We'd like to check that lightwalletd queries the mempool, but it looks like it doesn't do it after each GetMempoolTx request.
    //zebrad.expect_stdout_line_matches("answered mempool request req=TransactionIds")?;
```

**Comment by oxarbitrage** [comment]
I can't understand the CI error pointing to a file that does not exist: https://github.com/ZcashFoundation/zebra/runs/8173468346?check_suite_focus=true#step:13:654

I tried this PR locally and everything build and pass, new test passing.

File: zebrad/src/components/mempool.rs [hunk 3]
```diff
@@ -384,36 +405,68 @@ impl Service<Request> for Mempool {
     /// and will cause callers to disconnect from the remote peer.
     #[instrument(name = "mempool", skip(self, req))]
     fn call(&mut self, req: Request) -> Self::Future {
+        //let skip_full_validation = self.skip_full_validation();
+
```
Comment (oxarbitrage) [inline zebrad/src/components/mempool.rs]:
```suggestion
```

File: zebrad/tests/acceptance.rs [hunk 4]
```diff
@@ -102,6 +102,16 @@
...
 //! Please refer to the documentation of each test for more information.
+//!
+//! ## Disk Space for Testing
+//!
+//! The full sync and lightwalletd tests with cached state expect a temporary directory with
+//! at least 300 GB of disk space (2 copies of the full chain). To use another disk for the
+//! temporary test files:
+//!
+//! ```sh
+//! export TMPDIR=/path/to/disk/directory
```
Comment (oxarbitrage) [inline zebrad/tests/acceptance.rs:113]: nice!

File: zebrad/tests/common/launch.rs [hunk 5]
```diff
@@ -44,27 +47,14 @@ pub const BETWEEN_NODES_DELAY: Duration = Duration::from_secs(2);
 
 /// The amount of time we wait for lightwalletd to update to the tip.
 ///
-/// `lightwalletd` takes about 90 minutes to fully sync,
-/// and `zebrad` takes about 30 minutes to update to the tip.
-///
-/// TODO: reduce to 20 minutes when `zebrad` sync performance improves
-pub const LIGHTWALLETD_UPDATE_TIP_DELAY: Duration = Duration::from_secs(11 * 60 * 60);
+/// `lightwalletd` takes about 30-60 minutes to fully sync,
```
Comment (oxarbitrage) [inline zebrad/tests/common/launch.rs]:
```suggestion
/// `lightwalletd` takes about 60-120 minutes to fully sync,
```

**Review by oxarbitrage — DISMISSED** [review]
This looks really good, thank you for wrapping all this up. It costed me a lot to figure out what was going on here.

File: zebrad/src/components/mempool.rs (hunk 3 above)
Comment (teor2345) [inline zebrad/src/components/mempool.rs]:
```suggestion
```

Comment (teor2345) [inline zebrad/src/components/mempool.rs]:
```suggestion
```

**Comment by teor2345** [comment]
> I can't understand the CI error pointing to a file that does not exist: https://github.com/ZcashFoundation/zebra/runs/8173468346?check_suite_focus=true#step:13:654
> 
> I tried this PR locally and everything build and pass, new test passing.

It happened because two PRs changed the rpc config at the same time, so I needed to resolve the merge conflict in this PR. Until the merge conflict was resolved, GitHub was building the code from the original PR, or maybe building some kind of half-merged branch.

**Comment by teor2345** [comment]
This is #5069:
> kex_exchange_identification: Connection closed by remote host

https://github.com/ZcashFoundation/zebra/runs/8181572127?check_suite_focus=true#step:6:105

**Comment by teor2345** [comment]
I fixed a test bug by adding transactions to the send transactions list, rather than replacing the whole list with each new block.

**Comment by teor2345** [comment]
Oops, I forgot to push the latest fixes here.

**Comment by teor2345** [comment]
The beta Rust test failed, I opened PR #5090 to disable it.

We can merge this PR anyway, that test isn't required.

**Review by oxarbitrage — APPROVED** [review]

**Pull request merged**